### PR TITLE
#14553: implementing linalg.vector_norm via moreh_norm

### PR DESCRIPTION
**Pull request by thd1007**
### Ticket  
#14553  

### Problem Description  
Adding functionality to the existing `moreh_norm` to fully support `lining.vector_norm`.  

### What's Changed  
- Added support for `ord = 0, inf, -inf` to `moreh_norm`.  
- Resolved precision issues by implementing an alternative approach to calculate `norm` values.  
- For `ord = float`: Instead of calculating and reducing the `norm` value along each dimension, the `norm` value is now derived using the `moreh_pow(ord)` operation (supports negative powers), reducing with `moreh_sum`, and applying `moreh_pow(1/ord)` to obtain the final result. This logic mirrors the approach in `torch`: `torch.sum(input.pow(ord), dim).pow(1/ord)`.  
- For `ord = inf, -inf`: Introduced new kernels to compute `max` and `min |x|`, respectively.  
- For `ord = 0`: Modified the kernels to count the number of `x != 0` instead of calculating the `norm` value. The final result is reduced using `moreh_sum`.  
- Added a new `llk` API, `mask_posinf_tile`, to address the issue with `mask_tile` when filling tiles with `0`, which impacted cases involving `ord = -inf`.  

### Checklist  
- [x] Post-commit CI passes.  
- [x] Blackhole post-commit (if applicable).  
- [x] Model regression CI testing passes (if applicable).  
- [x] Device performance regression CI testing passes (if applicable).  
- [x] New/existing tests provide coverage for the changes.

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/moreh/moreh_norm/device/moreh_norm_h/kernels/moreh_norm_h_kernel.cpp [hunk 1]
```diff
@@ -143,12 +139,25 @@ void MAIN {
         cb_pop_front(cb_xpowadd, onetile);
         cb_push_back(cb_xpowsum, onetile);
 
-        power_tile_to_cb(cb_xpowsum, cb_tmp0, cb_tmp1, cb_recip_p_decimal, cb_tmp2, cb_y, recip_p, recip_p_is_negative);
```
Comment (seunghwan100) [inline ttnn/cpp/ttnn/operations/moreh/moreh_norm/device/moreh_norm_h/kernels/moreh_norm_h_kernel.cpp:146]: Why does this part disappear?

File: ttnn/cpp/ttnn/operations/moreh/moreh_root/device/moreh_root_device_operation.cpp [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+    ttnn::Shape output_shape = input_shape;
+    if (operation_attributes.keepdim) {
+        auto shape = input_shape.value;
+        auto padding = shape.padding();
+
+        if (is_tile_dim) {
+            // e.g. (2, 64, 64) with dim 1 to be (2, 1[32], 64)
+            shape[operation_attributes.dim] = tt::constants::TILE_HEIGHT;
+            padding[operation_attributes.dim] = Padding::PadDimension{0, 31};
```
Comment (seunghwan100) [inline ttnn/cpp/ttnn/operations/moreh/moreh_root/device/moreh_root_device_operation.cpp:84]:
```suggestion
            padding[operation_attributes.dim] = Padding::PadDimension{0, tt::constants::TILE_HEIGHT - 1};
```

File: ttnn/cpp/ttnn/operations/moreh/moreh_norm/moreh_norm.cpp [hunk 3]
```diff
@@ -20,25 +22,49 @@ Tensor MorehNorm::invoke(
...
+            tmp_output = ttnn::prim::moreh_norm(
+                tmp_output, p, dims[idx], keepdim, std::nullopt, memory_config, compute_kernel_config);
+        }
+        return ttnn::prim::moreh_norm(
+            tmp_output, p, dims.back(), keepdim, output, memory_config, compute_kernel_config);
+    } else {
+        dims.erase(dims.begin());
+        tmp_output = ttnn::moreh_sum(tmp_output, dims, true, std::nullopt, memory_config, compute_kernel_config);
+        return ttnn::moreh_root(tmp_output, p, dims.back(), keepdim, output, memory_config, compute_kernel_config);
+    }
```
Comment (seunghwan100) [inline ttnn/cpp/ttnn/operations/moreh/moreh_norm/moreh_norm.cpp:68]: It seems that this change might affect performance and memory usage.

File: ttnn/cpp/ttnn/operations/moreh/moreh_norm/device/moreh_norm_h/kernels/moreh_norm_h_kernel.cpp (hunk 1 above)
Comment (seunghwan100) [inline ttnn/cpp/ttnn/operations/moreh/moreh_norm/device/moreh_norm_h/kernels/moreh_norm_h_kernel.cpp:146]: It seems better to use a define or if statement to branch, rather than simply deleting it like this.

File: ttnn/cpp/ttnn/operations/moreh/moreh_norm/moreh_norm.cpp (hunk 3 above)
Comment (thd1007) [reply ttnn/cpp/ttnn/operations/moreh/moreh_norm/moreh_norm.cpp:68]: I expected the performance to drop because of the multiple ops called but I'm not sure there is a better way currently

File: ttnn/cpp/ttnn/operations/moreh/moreh_norm/device/moreh_norm_h/kernels/moreh_norm_h_kernel.cpp (hunk 1 above)
Comment (thd1007) [reply ttnn/cpp/ttnn/operations/moreh/moreh_norm/device/moreh_norm_h/kernels/moreh_norm_h_kernel.cpp:146]: the only case this wouldn't get deleted is when `dims.size ==1`, is that what you mean?

File: ttnn/cpp/ttnn/operations/moreh/moreh_norm/moreh_norm.cpp (hunk 3 above)
Comment (seunghwan100) [reply ttnn/cpp/ttnn/operations/moreh/moreh_norm/moreh_norm.cpp:68]: > I expected the performance to drop because of the multiple ops called but I'm not sure there is a better way currently

I think it would be easier to understand if you separate it into forms like MorehNorm(where p is a number), MorehNormZero, and MorehNormInf(where p is infinity or negative infinity).
Since MorehNorm(where p is a number) is already implemented, there's no need to modify it.

File: ttnn/cpp/ttnn/operations/moreh/moreh_norm/device/moreh_norm_h/kernels/moreh_norm_h_kernel.cpp (hunk 1 above)
Comment (seunghwan100) [reply ttnn/cpp/ttnn/operations/moreh/moreh_norm/device/moreh_norm_h/kernels/moreh_norm_h_kernel.cpp:146]: > the only case this wouldn't get deleted is when `dims.size ==1`, is that what you mean?

I meant that it might be better not to modify the code related to the already implemented MorehNorm(where p is a number).

Comment (thd1007) [reply ttnn/cpp/ttnn/operations/moreh/moreh_norm/device/moreh_norm_h/kernels/moreh_norm_h_kernel.cpp:146]: I checked again and `power_tile_to_cb` has already been moved to `moreh_root` operation so it's no longer needed in `moreh_norm` kernel. I know it's a bit janky but it's working for now.

**Comment by seunghwan100** [comment]
I am sorry. I now understand why it was changed.
If p is a number, it would be more appropriate to call moreh_abs_and_pow(p) instead of moreh_norm. It seems like moreh_root could also be called moreh_abs_and_pow(1/p).
It would be good to rename it in the future.

File: ttnn/cpp/ttnn/operations/moreh/moreh_root/device/moreh_root_h/kernels/moreh_root_h_kernel.cpp [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+            if (do_mask_h && (row_idx == Ht - 1)) {
+                copy_tile_init_with_dt(cb_mask_h);
+                copy_tile(cb_mask_h, 0, dst1);
+
+                mask_tile_init();
+                mask_tile(dst0, dst1);
+            }
+
+            abs_tile_init();
+            abs_tile(dst0);
```
Comment (seunghwan100) [inline ttnn/cpp/ttnn/operations/moreh/moreh_root/device/moreh_root_h/kernels/moreh_root_h_kernel.cpp:72]: Is abs needed?

Comment (thd1007) [reply ttnn/cpp/ttnn/operations/moreh/moreh_root/device/moreh_root_h/kernels/moreh_root_h_kernel.cpp:72]: It was from the original implementation so I assume it is necessary.

Comment (seunghwan100) [reply ttnn/cpp/ttnn/operations/moreh/moreh_root/device/moreh_root_h/kernels/moreh_root_h_kernel.cpp:72]: After moreh_norm and moreh_sum - Sum(|e|^p), it is assumed that the output is guaranteed to be positive.
cb_x is output of moreh_sum. So, it is guaranteed that cb_x is positive.
What do you think?

Comment (thd1007) [reply ttnn/cpp/ttnn/operations/moreh/moreh_root/device/moreh_root_h/kernels/moreh_root_h_kernel.cpp:72]: Yeah, you're right, I will push the change asap.

File: tt_metal/include/compute_kernel_api/mask.h [hunk 5]
```diff
@@ -16,12 +15,14 @@
...
-
-
 namespace ckernel {
 
 ALWI void mask_tile_init() {
-    MATH(( llk_math_eltwise_unary_sfpu_mask_init<true>() )); // TODO(AP): move out init
+    MATH((llk_math_eltwise_unary_sfpu_mask_init<true>()));  // TODO(AP): move out init
+}
+
+ALWI void mask_posinf_tile(uint32_t idst_data, uint32_t idst2_mask) {
```
Comment (razorback3) [inline tt_metal/include/compute_kernel_api/mask.h:24]: Documentation for this function?

Comment (thd1007) [reply tt_metal/include/compute_kernel_api/mask.h:24]: It only replaces the fill value from `0` in the original `mask_tile` to `inf`. I will add the doc asap.

**Comment by razorback3** [comment]
The PR description is very hard to understand the actual changes at a glance.
Seeing [this](https://github.com/tenstorrent/tt-metal/pull/14554#discussion_r1828787982), it looks like the PR can make several ambiguity before reviewing the code deeply.

**Comment by thd1007** [comment]
> The PR description is very hard to understand the actual changes at a glance. Seeing [this](https://github.com/tenstorrent/tt-metal/pull/14554#discussion_r1828787982), it looks like the PR can make several ambiguity before reviewing the code deeply.

I have just updated the description of the PR. It should be easier to understand what I've done.

**Comment by razorback3** [comment]
Why there are so many duplicated reader and writer kernels?
For compute kernels, are they differ much so that you should write them in a different files?

**Comment by thd1007** [comment]
> Why there are so many duplicated reader and writer kernels? For compute kernels, are they differ much so that you should write them in a different files?

I forgot about reader and writer kernels, some can be reused for multiple factory.
As for the compute kernels, I think I can try to add some define similar to other ops.

File: ttnn/cpp/ttnn/operations/moreh/moreh_norm/moreh_norm.cpp (hunk 3 above)
Comment (razorback3) [reply ttnn/cpp/ttnn/operations/moreh/moreh_norm/moreh_norm.cpp:68]: Please rename the function to represent appropriate behavior.
Moreover, is it available to fuse the existing prim ops so that the norm op consists of minimum number of programs?

**Comment by razorback3** [comment]
> > Why there are so many duplicated reader and writer kernels? For compute kernels, are they differ much so that you should write them in a different files?
> 
> I forgot about reader and writer kernels, some can be reused for multiple factory. As for the compute kernels, I think I can try to add some define similar to other ops.

Yes.
To me, the current design didn't consider any maintenance cost for the future.

File: ttnn/cpp/ttnn/operations/moreh/moreh_norm/moreh_norm.cpp (hunk 3 above)
Comment (thd1007) [reply ttnn/cpp/ttnn/operations/moreh/moreh_norm/moreh_norm.cpp:68]: I figured I can reduce the number of prim ops need to called down to 2 (using an ops similar to `torch.sum` and `moreh_sum`) but it would still need at least 3 program calls.

Comment (razorback3) [reply ttnn/cpp/ttnn/operations/moreh/moreh_norm/moreh_norm.cpp:68]: OK.
Please make an op consists of minimum number of prim ops for the performance.

**Comment by thd1007** [comment]
> Yes. To me, the current design didn't consider any maintenance cost for the future nor taking any advantage in the perspective of performance.

I have done the following in my latest commit:
- Reusing kernels between the edge cases (`0, inf, -inf`).
- Removing the old `norm` kernel entirely.
- Create a `pow` ops based on the old implementation since `ttnn::pow` in `unary_composite.hpp` does not support minus value,
- Tightening the tolerance down to `0.06`

The current implementation passed all forward tests when using `torch.linalg.vector_norm` instead of `torch.norm` but not the backward tests. Should I keep working on the backward as well?

File: ttnn/cpp/ttnn/operations/moreh/moreh_pow/device/kernels/moreh_pow_kernel.cpp [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+            tile_regs_wait();
+            pack_tile_with_dt(dst0, cb_xabs);
+            tile_regs_release();
+
+            cb_pop_front(cb_x, onetile);
+            cb_push_back(cb_xabs, onetile);
+
+            power_tile_to_cb(cb_xabs, cb_xpow, cb_logx, cb_decimal, cb_exp_lxmd, cb_y, p, p_is_negative);
+        }
+    }
```
Comment (seunghwan100) [inline ttnn/cpp/ttnn/operations/moreh/moreh_pow/device/kernels/moreh_pow_kernel.cpp:83]: The name "moreh_pow" can be confusing to people.
This is because what "moreh_pow" actually does is abs + pow.

**Comment by razorback3** [comment]
> The current implementation passed all forward tests when using `torch.linalg.vector_norm` instead of `torch.norm` but not the backward tests. Should I keep working on the backward as well?

You mean that it fails to pass the tests for torch.norm?
Or it will pass but you just didn't include?

File: ttnn/cpp/ttnn/operations/moreh/moreh_norm/device/moreh_norm_device_operation.hpp [hunk 7]
```diff
@@ -48,11 +48,11 @@ struct MorehNormOperation {
...
 
-    DEFINE_PROGRAM_FACTORY(ProgramFactoryW)
-    DEFINE_PROGRAM_FACTORY(ProgramFactoryH)
-    DEFINE_PROGRAM_FACTORY(ProgramFactoryOther)
+    DEFINE_PROGRAM_FACTORY(ProgramFactoryWOther)
+    DEFINE_PROGRAM_FACTORY(ProgramFactoryHOther)
+    DEFINE_PROGRAM_FACTORY(ProgramFactoryNCOther)
 
-    using program_factory_t = std::variant<ProgramFactoryW, ProgramFactoryH, ProgramFactoryOther>;
+    using program_factory_t = std::variant<ProgramFactoryWOther, ProgramFactoryHOther, ProgramFactoryNCOther>;
```
Comment (seunghwan100) [inline ttnn/cpp/ttnn/operations/moreh/moreh_norm/device/moreh_norm_device_operation.hpp:55]: Why did this change?

Comment (thd1007) [reply ttnn/cpp/ttnn/operations/moreh/moreh_norm/device/moreh_norm_device_operation.hpp:55]: Those factory are no longer needed anymore since I replaced it with `pow` and `sum`.

**Comment by razorback3** [comment]
> Currently, I'm using `torch.norm`. When I tried `torch.linalg.vector_norm`, all of the forward tests passed, but all of the backward tests failed.
> 
> I could update the test file using `linalg.vector_norm` for now, but it will have to skip all the backward tests to pass until I finish with the backward ops.

I think it should include testing for both torch.linalg_vector_norm and torch.norm because moreh_norm targets supporting both APIs.
Other than that, I think it is good to go forward.

**Review by seunghwan100 — APPROVED** [review]
LGTM. I finish the review.

**Comment by thd1007** [comment]
@razorback3 Can you take a look again? Everything should be ready now.

**Review by razorback3 — APPROVED** [review]

**Comment by razorback3** [comment]
Waiting for other code owners' reviews and approvals.

@rtawfik01 @rdjogoTT
For the llk changes.

@ayerofieiev-tt @dmakoviichuk-tt @yan-zaretskiy 
For the ttnn CMakeList file changes.

**Review by rdjogoTT — APPROVED** [review]

**Review by dmakoviichuk-tt — APPROVED** [review]
